src/lib/protocols/quic.c

```diff
@@ -31,107 +31,135 @@
-      NDPI_LOG(NDPI_PROTOCOL_QUIC, ndpi_struct, NDPI_LOG_DEBUG, "calculating quic over udp.\n");
-
-      if((((sport == 80) || (dport == 80) || (sport == 443) || (dport == 443))))
-      {
-	NDPI_LOG(NDPI_PROTOCOL_QUIC, ndpi_struct, NDPI_LOG_DEBUG, "exclude quic.\n");
-	NDPI_ADD_PROTOCOL_TO_BITMASK(flow->excluded_protocol_bitmask, NDPI_PROTOCOL_QUIC);
-
-        if (packet->payload[0] & QUIC_VER_MASK)
-	{
-	  // Skip CID length.
-	  ver_offs = connect_id(packet->payload[0]);
-
-	  if (ver_offs >= 0)
-	  {
-	    unsigned char vers[] = {packet->payload[ver_offs], packet->payload[ver_offs + 1],
-				    packet->payload[ver_offs + 2], packet->payload[ver_offs + 3]};
-
-	    // Version Match.
-	    if ((vers[0] == 'Q' && vers[1] == '0') &&
-		((vers[2] == '3' && (vers[3] == '3' || vers[3] == '2' || vers[3] == '1' || vers[3] == '0' )) ||
-		 (vers[2] == '2' && (vers[3] == '9' || vers[3] == '8' || vers[3] == '7' || vers[3] == '6' ||
-		                     vers[3] == '5' || vers[3] == '4' || vers[3] == '3' || vers[3] == '2' ||
-				     vers[3] == '1' || vers[3] == '0')) ||
-		 (vers[2] == '1' && (vers[3] == '9' || vers[3] == '8' || vers[3] == '7' || vers[3] == '6' ||
-				     vers[3] == '5' || vers[3] == '4' || vers[3] == '3' || vers[3] == '2' ||
-				     vers[3] == '1' || vers[3] == '0')) ||
-		 (vers[2] == '0' && vers[3] == '9')))
-
-	    {
-	      NDPI_LOG(NDPI_PROTOCOL_QUIC, ndpi_struct, NDPI_LOG_DEBUG, "found quic.\n");
-	      ndpi_int_quic_add_connection(ndpi_struct, flow);
-	    }
-	  }
-	}
-      }
-      else
-      {
-	NDPI_LOG(NDPI_PROTOCOL_QUIC, ndpi_struct, NDPI_LOG_DEBUG, "exclude quic.\n");
-	NDPI_ADD_PROTOCOL_TO_BITMASK(flow->excluded_protocol_bitmask, NDPI_PROTOCOL_QUIC);
-      }
+    // UDP 2-bytes F/P
+    if(packet->udp != NULL && packet->payload_packet_len > 2)
+    {
+    	u_int16_t sport = ntohs(packet->udp->source), dport = ntohs(packet->udp->dest);
+
+    	NDPI_LOG(NDPI_PROTOCOL_QUIC, ndpi_struct, NDPI_LOG_DEBUG, "calculating QUIC over udp.\n");
+
+        if (sport == 80 || dport == 80 || sport == 443 || dport == 443)
+        {
+            // Check if NO_VERS. PUBLIC FLAGS & SEQ/CID MASK == 0x0. SEQ must be -gt 0.
+            if ((packet->payload[0] & QUIC_NO_V_RES_RSV) == 0 && sequence(packet->payload))
+            {
+                NDPI_LOG(NDPI_PROTOCOL_QUIC, ndpi_struct, NDPI_LOG_DEBUG, "found QUIC.\n");
+                ndpi_int_quic_add_connection(ndpi_struct, flow);
+            }
+
+            // Version presence; CID length
+            if (packet->payload[0] & QUIC_VER_MASK)
+            {
+                // Skip CID length
+                ver_offs = connect_id(packet->payload[0]);
+
+                if (ver_offs >= 0)
+                {
+                    const uint8_t *vers = &packet->payload[ver_offs];
+                    short int res = DIGIT(vers[1], vers[2], vers[3]);
+
+                    if (*vers == 'Q')
+                    {
+                        // QUIC versions. '22', '28' < omitted.
```

Comment:
I did not see the presence of these versions but if you can confirm, can be added.